File: alpha/experiments/item-storage-protocol-standalone/src/exp/e0_identity.ts
// E0. Identity — both parties exist as keys, nothing more.
// We generate the two keypairs (done in World), derive each identifier from the
// public key, pin each other's key, and prove that recognition (signature check)
// and counting (identifier derivation) rest on the same public key.

import type { World } from "../world.ts";
import { Checker, type ExperimentResult } from "../types.ts";
import { Actor } from "../actor.ts";
import { signMessage, verifyMessage } from "../crypto.ts";

export function run(world: World): ExperimentResult {
  const c = new Checker();
  const { customer, provider } = world;

  // Each actor records a self-describing genesis entry (its id, its public key,
  // and the peer key it has pinned), signed by itself.
  const ts = world.clock.iso();
  customer.ledger.append("genesis", ts, {
    self: customer.id, publicKey: customer.publicKeyHex, pinnedPeer: provider.id,
  }, [customer.signer()]);
  provider.ledger.append("genesis", ts, {
    self: provider.id, publicKey: provider.publicKeyHex, pinnedPeer: customer.id,
  }, [provider.signer()]);

  // A message signed by the customer verifies under the customer's pinned key...
  const msg = "I, Ada, am bringing my items to the co-op.";
  const sig = signMessage(customer.keypair, msg);
  c.ok(
    "customer signature verifies under customer's pinned key",
    verifyMessage(world.keyring[customer.id], msg, sig),
  );
  // ...and fails under the provider's key. Adversarial: a forged attribution.
  c.ok(
    "customer signature does NOT verify under provider's key",
    !verifyMessage(world.keyring[provider.id], msg, sig),
  );

  // Identifier derivation is a deterministic function of the public key alone.
  c.ok(
    "identifier derivation is deterministic",
    Actor.deriveId(customer.publicKeyHex) === customer.id &&
      Actor.deriveId(provider.publicKeyHex) === provider.id,
  );
  // Distinct keys yield distinct identifiers.
  c.ok("distinct keys yield distinct identifiers", customer.id !== provider.id);

  return {
    id: "E0",
    title: "Identity",
    plainSentence: "We recognize you the same way we count you.",
    assertions: c.results,
    tables: [
      {
        title: "Pinned identities",
        headers: ["actor", "id", "public key (hex, truncated)"],
        rows: [
          ["Ada (customer)", customer.id, customer.publicKeyHex.slice(0, 24) + "…"],
          ["Co-op (provider)", provider.id, provider.publicKeyHex.slice(0, 24) + "…"],
        ],
      },
    ],
    notes: [],
  };
}

File: alpha/experiments/item-store/tests/e0_identity.rs
//! Phase 1 wiring test — ports E0 (identity) from the standalone oracle
//! (`src/exp/e0_identity.ts`): recognition (signature verification) and counting
//! (identifier derivation) rest on the same public key.
//!
//! This is the RED→GREEN gate for Phase 1: it exercises the crate's public API
//! (derive keypair → sign → verify under a pinned key → derive id), not an
//! isolated internal.

use item_store::crypto::{derive_keypair, public_key_from_hex, verify_message};
use item_store::identity::derive_id;

const MASTER_SEED: &str = "item-store::e0::test-seed";

/// Ports E0's four assertions plus a tamper edge:
/// 1. a message signed by the customer verifies under the customer's pinned key;
/// 2. it does NOT verify under the provider's key (a forged attribution);
/// 3. identifier derivation is a deterministic function of the public key;
/// 4. distinct keys yield distinct identifiers;
/// 5. (edge) a tampered message does not verify under the true key.
#[test]
fn e0_identity_recognize_and_count() {
    let customer = derive_keypair(MASTER_SEED, "customer");
    let provider = derive_keypair(MASTER_SEED, "provider");

    let customer_id = derive_id(&customer.verifying_key());
    let provider_id = derive_id(&provider.verifying_key());

    // Each party pins the other's public key by reconstructing it from the
    // published hex — the same path a networked peer would take.
    let customer_pin =
        public_key_from_hex(&customer.public_key_hex()).expect("customer key is valid hex");
    let provider_pin =
        public_key_from_hex(&provider.public_key_hex()).expect("provider key is valid hex");

    let message = "I, Ada, am bringing my items to the co-op.";
    let signature = customer.sign_message(message);

    // 1. verifies under the customer's pinned key
    assert!(
        verify_message(&customer_pin, message, &signature),
        "customer signature verifies under customer's pinned key",
    );

    // 2. does NOT verify under the provider's key (adversarial: forged attribution)
    assert!(
        !verify_message(&provider_pin, message, &signature),
        "customer signature must not verify under provider's key",
    );

    // 3. identifier derivation is deterministic in the public key alone —
    //    deriving from the reconstructed (pinned) key reproduces the id.
    assert_eq!(
        derive_id(&customer_pin),
        customer_id,
        "identifier derivation is deterministic (customer)",
    );
    assert_eq!(
        derive_id(&provider_pin),
        provider_id,
        "identifier derivation is deterministic (provider)",
    );

    // 4. distinct keys yield distinct identifiers
    assert_ne!(
        customer_id, provider_id,
        "distinct keys yield distinct identifiers",
    );

    // 5. (edge) tampering with the message breaks verification under the true key
    let tampered = "I, Ada, am bringing my items to the co-op!";
    assert!(
        !verify_message(&customer_pin, tampered, &signature),
        "a tampered message must not verify under the true key",
    );
}
